sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsGdsStore.scala

```diff
@@ -39,30 +47,75 @@ class RapidsGdsStore(
         case d: DeviceMemoryBuffer => d
         case _ => throw new IllegalStateException("copying from buffer without device memory")
       }
-      val id = other.id
-      val path = if (id.canShareDiskPaths) {
-        sharedBufferFiles.computeIfAbsent(id, _ => id.getDiskPath(diskBlockManager))
+      if (isBatched(deviceBuffer.getLength)) {
+        batchSpill(other, deviceBuffer)
       } else {
-        id.getDiskPath(diskBlockManager)
+        singleShotSpill(other, deviceBuffer)
       }
-      // When sharing files, append to the file; otherwise, write from the beginning.
-      val fileOffset = if (id.canShareDiskPaths) {
-        // only one writer at a time for now when using shared files
-        path.synchronized {
-          CuFile.appendDeviceBufferToFile(path, deviceBuffer)
+    }
+  }
+
+  private def isBatched(length: Long): Boolean = {
+    length < BatchWriteBufferSize
+  }
+
+  private def singleShotSpill(
+      other: RapidsBuffer, deviceBuffer: DeviceMemoryBuffer): RapidsBufferBase = {
+    val id = other.id
+    val path = if (id.canShareDiskPaths) {
+      sharedBufferFiles.computeIfAbsent(id, _ => id.getDiskPath(diskBlockManager))
+    } else {
+      id.getDiskPath(diskBlockManager)
+    }
+    // When sharing files, append to the file; otherwise, write from the beginning.
+    val fileOffset = if (id.canShareDiskPaths) {
+      // only one writer at a time for now when using shared files
+      path.synchronized {
+        CuFile.appendDeviceBufferToFile(path, deviceBuffer)
+      }
+    } else {
+      CuFile.writeDeviceBufferToFile(path, 0, deviceBuffer)
+      0
+    }
+    logDebug(s"Spilled to $path $fileOffset:${other.size} via GDS")
+    new RapidsGdsBuffer(id, None, fileOffset, other.size, other.meta, other.getSpillPriority,
+      other.spillCallback)
+  }
+
+  private def batchSpill(
+      other: RapidsBuffer, deviceBuffer: DeviceMemoryBuffer): RapidsBufferBase = this.synchronized {
+      if (deviceBuffer.getLength > BatchWriteBufferSize - batchWriteBufferOffset) {
+        val path = batchWriteBufferId.getDiskPath(diskBlockManager).getAbsolutePath
+        withResource(new CuFileWriteHandle(path)) { handle =>
+          handle.write(batchWriteBuffer, batchWriteBufferOffset, 0)
+          logDebug(s"Spilled to $path 0:$batchWriteBufferOffset via GDS")
         }
-      } else {
-        CuFile.writeDeviceBufferToFile(path, 0, deviceBuffer)
-        0
+        pendingBuffers.foreach(_.cuFileBuffer = None)
```

Comment:
I'm confused by this line.  `CuFileBuffer` resources need to be explicitly closed, we can't just overwrite the references with `None`.  To me this seems like a clear leak or if this is being closed earlier then the `None` overwrite should occur there rather than here.

Reply:
Wondering if it would be clearer and more consistent if this did use reference-counting, otherwise we've got some code that needs to close a cufilebuffer and other code that _must not_ close the cufile buffer, and we may confuse which is which.  If we always close when we're done then it could be simpler.